### PR TITLE
Sparse alpha

### LINKED ISSUE #42: Alpha diversity operates on dense feature tables
**Bug Description**
Alpha diversity calculations are occurring on dense feature tables. These transformations require >100x resources than what is necessary as it is not unusual for microbiome data to be < 1% dense. I believe it is appropriate to remark as a regression bug as Q1 operated on sparse data.

**Expected behavior**
Alpha diversity should be computed using memory proportional to the number of nonzero entries, rather than to the number of samples and features. 

**Comments**
A non-perfect, but appropriate and quick to do work around, would be to compute alpha per-sample by iterating over a BIOM table with `.iter()`, requesting `dense=True`, and computing alpha diversity on each `np.array`. If needed, this could be parallelized through joblib. This is a much lower burden change than adapting `skbio`'s alpha diversity call to properly support sparse vectors (and which may be complex for some metrics).

I can issue a PR soon on this.

### PR DESCRIPTION
**Pull request by wasade**
Fixes #42. Changes operations to rely on the `biom.Table` API, and perform alpha calculations on a per-sample basis which avoids a full conversion to a dense matrix.

### CODE REVIEW

**Comment by wasade** [comment]
It is very likely that all the `alpha_diversity` calls should receive `validate=False` as many of the checks are extraneous. We can do a bulk check for negative or nan upstream as well if necessary.

**Comment by wasade** [comment]
...it seems `validate=False` still performs some validation, and the exceptions suggest an upstream transform is needed. Will evaluate in a little bit

**Comment by wasade** [comment]
Test failures seem like they may be upstream

**Comment by wasade** [comment]
The table validation has been updated to check for NaN. It wasn't updated to check for negative values as I worry that may be too restrictive, as it would prohibit application to some types of transformed data.

**Comment by cherman2** [comment]
Hey @wasade, Are you still working on this issue?

**Comment by wasade** [comment]
The failures were upstream? It's been > 18 months since I've looked at this so not sure if the upstream issues were sorted out

**Comment by wasade** [comment]
Is there a reason this wasn't merged previously?

**Comment by ebolyen** [comment]
I don't recall there being any reason beyond losing track of it. Perhaps CI was just failing for no particularly interesting reason?

I think if we want to move forward with this, we need to rebase on dev and check that everything still works.

**Comment by ebolyen** [comment]
(For context, we've been going through all of our issues and finding stragglers which were lost from our project boards at certain points.)

**Comment by wasade** [comment]
Sounds good, thanks @ebolyen. Can that be done w/o me?

**Comment by wasade** [comment]
I think that addresses the conflicts. It would be really good to revert the system calls to unifrac back to the API calls but that's out of scope here.

**Comment by wasade** [comment]
@ebolyen @cherman2 I think the lingering issue is upstream. It would be really nice to get this merged.

**Comment by wasade** [comment]
Just checking in on this, any traction on the upstream issues?

File: q2_diversity_lib/alpha.py [hunk 1]
```diff
@@ -44,70 +44,76 @@
 
 
 # --------------------- Phylogenetic -----------------------------------------
-@_disallow_empty_tables
+@_validate_tables
 @_validate_requested_cpus
```
Comment (gregcaporaso) [inline q2_diversity_lib/alpha.py:49]:
```suggestion
```

This ~~seems to~~ may be the issue - the `threads` parameter was removed, but the call to `_validate_requested_cpus` is still happening.

Comment (wasade) [reply q2_diversity_lib/alpha.py:49]: Cool thank you!

Comment (gregcaporaso) [reply q2_diversity_lib/alpha.py:49]: Ok, that doesn't seem to have been it unfortunately.

**Comment by gregcaporaso** [comment]
The [action registration](https://github.com/qiime2/q2-diversity-lib/blob/ac25d75f1e409702e25bb14dd228c05a324e1678/q2_diversity_lib/plugin_setup.py#L59) wasn't updated, so the parameters don't align between the QIIME 2 action and the underlying function. But that said, this introduces an interface change which we avoid without giving notice. Is removing the `threads` parameter critical for this?

**Comment by wasade** [comment]
Oh good spot. Let me revert that, the `faith_pd` function in `unifrac` doesn't explicitly support `threads` yet, and I think this PR came in before we discovered the wide threading triggered by not setting a value for `OMP_NUM_THREADS`. I likely can't do it today but will tomorrow. 

One thing, separate from this PR, that we should do is update the UniFrac calls themselves which avoid intermediate IO for tables / trees / distances via zero-copy mechanisms.

**Comment by gregcaporaso** [comment]
@wasade, do you happen to know if there is a way that we can test that the number of threads doesn't exceed expectations - for example using [threading.active_count](https://docs.python.org/3/library/threading.html#threading.active_count) (though I don't know if that's relevant for how the threads are being created in UniFrac). Your mention of not setting `OMP_NUM_THREADS` got me thinking about that again.

**Comment by wasade** [comment]
`OMP_NUM_THREADS` is best managed centrally in QIIME 2, if it isn't already, as OpenMP is a very common framework. 

For the failure here, we need to revert to the cmd wrapper that was previously in place as the `faith_pd` function call does not currently expose a parameter for threads in the Python API.

**Comment by wasade** [comment]
@gregcaporaso, I think this is good to go!

**Review by gregcaporaso — APPROVED** [review]
Ok, agreed that this looks good now. Thanks @wasade.

**Comment by wasade** [comment]
Thank you, @gregcaporaso!

**Pull request merged**